mock/src/lib.rs

```diff
@@ -0,0 +1,349 @@
+
+    /// Returns the config used to create this instance.
+    pub fn config(&self) -> Config {
+        self.inner.borrow().config().clone()
+    }
+
+    fn new(event_tx: Sender<Event>, config: Config) -> Self {
+        Self {
+            inner: Node::new(event_tx, config),
+        }
+    }
+}
+
+#[cfg(test)]
+impl QuicP2p {
+    fn addr(&self) -> SocketAddr {
+        *self.inner.borrow().addr()
+    }
+
+    fn our_type(&self) -> OurType {
+        self.inner.borrow().our_type()
+    }
+}
+
+/// Configuration for `QuicP2p`.
+#[derive(Debug, Default, Clone, Eq, PartialEq, Serialize, Deserialize, StructOpt)]
+pub struct Config {
+    /// Hard-coded contacts.
+    #[structopt(
+        short,
+        long,
+        default_value = "[]",
+        parse(try_from_str = "serde_json::from_str")
+    )]
+    pub hard_coded_contacts: HashSet<NodeInfo>,
+    /// Type of our `QuicP2p` instance: node or client.
+    #[structopt(short = "t", long, default_value = "node")]
+    pub our_type: OurType,
+    /// Port to listen to.
+    pub ip: Option<IpAddr>,
+    /// IP address to listen to.
+    pub port: Option<u16>,
+}
+
+impl Config {
+    /// Create `Config` for node.
+    pub fn node() -> Self {
+        Self {
+            our_type: OurType::Node,
+            ..Self::default()
+        }
+    }
+
+    /// Create `Config` for client.
+    pub fn client() -> Self {
+        Self {
+            our_type: OurType::Client,
+            ..Self::default()
+        }
+    }
+
+    /// Set the `hard_coded_contacts`.
+    pub fn with_hard_coded_contacts<I>(self, contacts: I) -> Self
+    where
+        I: IntoIterator,
+        I::Item: Into<NodeInfo>,
+    {
+        Self {
+            hard_coded_contacts: contacts.into_iter().map(Into::into).collect(),
+            ..self
+        }
+    }
+
+    /// Set the `hard_coded_contacts` to a single contact.
+    pub fn with_hard_coded_contact<T>(self, contact: T) -> Self
+    where
+        T: Into<NodeInfo>,
+    {
+        self.with_hard_coded_contacts(iter::once(contact))
+    }
+
+    /// Set the endpoint (IP + port) to use.
+    pub fn with_endpoint(self, addr: SocketAddr) -> Self {
+        Self {
+            ip: Some(addr.ip()),
+            port: Some(addr.port()),
+            ..self
+        }
+    }
+}
+
+/// The type of our `QuicP2p` instance: client or node.
+#[derive(Clone, Copy, Eq, PartialEq, Debug, Serialize, Deserialize)]
+pub enum OurType {
+    /// We are a client
+    Client,
+    /// We are a node
+    Node,
+}
+
+impl Default for OurType {
+    fn default() -> Self {
+        Self::Node
+    }
+}
+
+impl FromStr for OurType {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s.to_lowercase().as_str() {
+            "client" => Ok(Self::Client),
+            "node" => Ok(Self::Node),
+            x => Err(format!("Unknown client type: {}", x)),
+        }
+    }
+}
+
+/// Events from `QuicP2p` to the user.
+#[derive(Debug)]
+pub enum Event {
```

Comment:
Since it is in the same repo, is it worth using the real event?

Reply:
No, I guess since it does not have a dependency on quic_p2p, make sense.

Reply:
Well, currently mock doesn't depend on the real thing, but maybe it should? Then we would be able to reuse some of the definitions. That could also make it a bit easier to ensure both APIs remain identical. What do you think?

Reply:
>That could also make it a bit easier to ensure both APIs remain identical

This would make total sense, especially if we start to change the code to use async/await features. However, I think we can create an issue for now and tackle this later.